src/swapRouter.ts

```diff
@@ -302,14 +330,97 @@ export abstract class SwapRouter {
       }
     }
 
-    // refund
-    if (mustRefund) {
+    // must refund when paying in ETH, but with an uncertain input amount
+    if (inputIsNative && sampleTrade.tradeType === TradeType.EXACT_OUTPUT) {
       calldatas.push(Payments.encodeRefundETH())
     }
 
     return {
       calldata: MulticallExtended.encodeMulticall(calldatas, options.deadlineOrPreviousBlockhash),
-      value: toHex(totalValue.quotient),
+      value: toHex(inputIsNative ? totalAmountIn.quotient : ZERO),
     }
   }
+
+  /**
+   * Produces the on-chain method name to call and the hex encoded parameters to pass as arguments for a given trade.
+   * @param trade to produce call parameters for
+   * @param options options for the call parameters
+   */
+  public static swapAndAddCallParameters(
+    trades:
+      | Trade<Currency, Currency, TradeType>
+      | V2Trade<Currency, Currency, TradeType>
+      | V3Trade<Currency, Currency, TradeType>
+      | (V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>)[],
+    options: SwapOptions,
+    position: Position,
+    addLiquidityOptions: AddLiquidityOptions
+  ): MethodParameters {
+    const {
+      calldatas,
+      inputIsNative,
+      outputIsNative,
+      totalAmountIn: totalAmountSwapped,
+      totalAmountOut,
+    } = SwapRouter.encodeSwaps(trades, options, true)
+
+    const zeroForOne = position.pool.token0 === totalAmountSwapped.currency.wrapped
+    const { positionAmountIn, positionAmountOut } = SwapRouter.getPositionAmounts(position, zeroForOne)
+
+    // if tokens are native they will be converted to WETH9
+    const tokenIn = inputIsNative ? WETH9[1] : positionAmountIn.currency.wrapped
+    const tokenOut = outputIsNative ? WETH9[1] : positionAmountOut.currency.wrapped
+
+    // if swap output does not make up whole outputTokenBalanceDesired, pull in remaining tokens for adding liquidity
+    const amountOutRemaining = positionAmountOut.subtract(totalAmountOut.wrapped)
+    if (amountOutRemaining.greaterThan(CurrencyAmount.fromRawAmount(positionAmountOut.currency, 0))) {
```

Comment:
hmm, i guess this is for the case when the user already has some of the output currency, but it's not in the appropriate ratio with the input? so you need to either wrap or pull after the swap? guess that makes sense

Reply:
in this situation, the user is only getting a partial amount of a token from the swap, but already owns the rest.

Reply:
This is also why "single asset add" always confused me as a term. I always thought of it as being potentially 2 assets as well.